### PR TITLE
feat: Add Seed Support for Reproducible and Deterministic Generation in VoxCPM and VoxCPM2

### PR DESCRIPTION
**Pull request by elismasilva**
#### **Overview**
This Pull Request adds optional random seed (`seed`) support across the `nano-vllm-voxcpm` stack for both VoxCPM (v1/v1.5) and VoxCPM2.

The goal is reproducible speech generation when a caller explicitly provides a seed, while preserving the previous random behavior and fast batched path when no seed is provided.

---

##### **1. Request-Scoped Seed Propagation**
* Added `seed` support to the public generation paths for both VoxCPM and VoxCPM2.
* Propagated `seed` through the async/sync server pools, IPC server calls, engine `add_request()`, sequence payloads, and runner payloads.
* Avoided process-global RNG mutation in the request path. The implementation no longer relies on `torch.manual_seed(seed)` or `torch.cuda.manual_seed_all(seed)` in `add_request()`, so one request cannot overwrite the global RNG state used by another queued or batched request.

##### **2. Per-Step DiT Noise Without Repeating the Same Initial Noise**
* Added a request-local `seed_step` counter to the VoxCPM and VoxCPM2 sequence payloads.
* The runner derives a deterministic per-step seed via `derive_step_seed(seed, seed_step)` before generating seeded DiT `z_noise`.
* `postprocess_seq()` increments `seed_step` after each successfully generated latent, so the same request receives a deterministic RNG stream that advances across autoregressive decode steps.
* The same input and original seed still map to the same deterministic sequence of per-step noise values across runs, but each DiT call no longer restarts from identical initial noise.

##### **3. Seeded and Unseeded Noise Paths**
* Restored the common unseeded path to use batched noise generation:

```python
torch.randn((bsz, feat_dim, patch_size), ...)
```

* For mixed batches, the runner first generates the full batch normally and then overwrites only rows that belong to requests with a non-negative seed.
* For fully seeded batches, the runner allocates the batch tensor and fills each row from the request-local derived seed without consuming the process-global RNG stream.
* `z_noise` remains an explicit model input, so RNG generation happens outside the model forward/CUDA graph capture path.

##### **4. FastAPI REST Endpoint & Backward Compatibility**
* Added `seed: int | None = Field(None, ...)` to the deployment `GenerateRequest` schema.
* Updated `deployment/app/api/routes/generate.py` to inspect `server.generate` before forwarding optional compatibility-sensitive arguments.
* `seed` is only added to `generate_kwargs` when the client provides `req.seed`.
* If a client provides `seed` but the loaded server implementation does not accept a `seed` parameter, the route returns a clean `HTTPException(400)`.
* When no seed is supplied, the route does not pass `seed`, preserving the previous default/random behavior.

##### **5. Benchmarking Suite**
* Added `--seed` to:
  * `benchmark/bench_inference.py`
  * `benchmark/bench_open_loop_users.py`
  * `benchmark/bench_closed_loop_users.py`
* Propagated the seed through warmup and measured benchmark paths for both in-process and HTTP execution modes where applicable.

##### **6. Documentation**
* Documented benchmark `--seed` usage in `benchmark/README.md`.
* Documented the deployment API `seed` request field in `deployment/README.md`.

---

#### **Verification & Testing**

Targeted seed behavior tests:

```powershell
python -m pytest tests/unit/test_seed_utils.py tests/unit/test_voxcpm_engine_max_model_len.py::test_postprocess_advances_seed_step_after_generated_latent tests/unit/test_voxcpm2_engine_max_model_len.py::test_postprocess_advances_seed_step_after_generated_latent tests/unit/test_voxcpm2_chunk_sizes.py::test_voxcpm2_runner_slices_decoded_waveform_with_decoder_chunk_size -q
```

Focused regression set:

```powershell
python -m pytest tests/unit/test_seed_utils.py tests/unit/test_voxcpm_engine_max_model_len.py tests/unit/test_voxcpm2_engine_max_model_len.py tests/unit/test_voxcpm2_chunk_sizes.py -q
```

Result:

```text
18 passed
```

Syntax validation:

```powershell
python -m compileall nanovllm_voxcpm/utils/seed.py nanovllm_voxcpm/models/voxcpm/engine.py nanovllm_voxcpm/models/voxcpm/runner.py nanovllm_voxcpm/models/voxcpm2/engine.py nanovllm_voxcpm/models/voxcpm2/runner.py tests/unit/test_seed_utils.py tests/unit/test_voxcpm_engine_max_model_len.py tests/unit/test_voxcpm2_engine_max_model_len.py tests/unit/test_voxcpm2_chunk_sizes.py
```

Relevant commits:

```text
039d262 feat: add seed support for reproducible generation in both VoxCPM and VoxCPM2
f6300b7 refactor: secure per-request seeding under concurrency and fix API backward compatibility
461c351 Fix per-step seeded DiT noise
```

### CODE REVIEW

**Comment by a710128** [comment]
Thanks for adding seed support and wiring it through the benchmarks/deployment API — this will be very useful for reproducibility.

I noticed two areas that I think need a bit more work before merging:

1. **Per-request seed behavior under concurrency**

   The current implementation calls `torch.manual_seed(seed)` / `torch.cuda.manual_seed_all(seed)` in `add_request()`. Since this changes the process-global RNG state, it is not really request-scoped. The actual sampling happens later during `srv.step()`, so if multiple requests are queued/batched, a later request can overwrite the RNG state used by earlier requests. It may also affect requests that did not pass a seed.

   Since the seed mainly affects the DiT sampling part, could we look for a way to make the seed part of the per-request/sequence sampling state instead of changing the global RNG? For example, passing the seed down to the DiT/sampler path and using a per-request `torch.Generator` or equivalent would make reproducibility much safer in concurrent serving.

2. **Backward compatibility for the deployment `generate()` interface**

   The deployment route now always forwards `seed=req.seed` to `server.generate(...)`. That can break older/custom server implementations whose `generate()` method does not accept a `seed` keyword.

   Could we make this compatible with the old interface? One possible approach is to inspect the `generate()` signature before adding `seed` to `generate_kwargs`, similar to the existing `ref_audio_latents` compatibility check. Also, when the client does not provide `seed`, we should preserve the previous random behavior — i.e. do not force determinism, and let the server use a random seed / normal RNG path.

Overall the feature direction looks good; I think addressing these two points would make the seed support safer for real deployment and concurrent benchmark scenarios.

**Comment by elismasilva** [comment]
> Thanks for adding seed support and wiring it through the benchmarks/deployment API — this will be very useful for reproducibility.
>...

You were totally right to raise these points, as the first implementation did not take the distributed environment into account. 

To fix the global RNG overwrite issue under concurrent batching, I completely redesigned how the seed is handled. Instead of changing the global PyTorch seed in the server, the seed is now propagated through the sequence payloads down to the runner. In `runner.py`, we pre-allocate the initial noise tensor `z_noise` on the host (Python/CPU) and seed each row of the batch individually using its own `torch.Generator`. This `z_noise` is then passed as a static tensor input all the way into `UnifiedCFM.forward`. This avoids any `.item()` calls or dynamic loops inside the compiled forward pass, ensuring concurrent-safe determinism while keeping CUDA Graphs fully active and optimized on Linux.

For the deployment API, I updated `api/routes/generate.py` to follow the exact same design pattern you previously implemented for `ref_audio_latents`. It now uses `inspect.signature` to check if `seed` is supported by the loaded model before forwarding the argument, throwing a clean `HTTPException(400)` if a seed is provided but unsupported. It also gracefully preserves the default random behavior when no seed is supplied.

Since I don't have a multi-GPU distributed environment to simulate and test on, I would really appreciate it if you could validate the distributed/tensor-parallel execution on your end. That said, in all single-GPU tests on both Linux and Windows, it works successfully and the generated waveforms are identical.

Let me know if this looks good to you!

**Comment by a710128** [comment]
Thanks for the update — the API compatibility part looks good to me now.

I still have one concern about how the RNG is implemented for the DiT noise. The current approach creates a new `torch.Generator(...).manual_seed(seed)` inside each runner invocation. This avoids global RNG contamination, but it also means the same request receives the same initial DiT noise at every autoregressive decode step. In practice this can noticeably hurt generation quality, because each diffusion call starts from identical noise.

I think we need the per-request RNG stream to advance across decode steps. A couple of possible approaches:

- Keep a request-local RNG state / `torch.Generator` for the lifetime of the sequence, and use it outside the CUDA graph to create `z_noise` each step. `z_noise` can still be passed into the CUDA graph as an input tensor, so random generation itself does not need to be captured.
- If storing a `torch.Generator` is inconvenient, keep a request-local `seed_step` counter and derive a deterministic but different seed per DiT call, e.g. from `(seed, seed_step)`, then increment it after each successful generated latent.

Also, could we keep the seeded and unseeded paths separate?

For the common unseeded path, it would be better to preserve the previous batched `torch.randn((bsz, ...))` behavior. The current per-sequence loop may regress throughput even when no request provides a seed. For mixed batches, one possible pattern is to generate the full batch noise normally, then only overwrite the rows that have per-request seeds.

So the main request is: seeded requests should be deterministic and request-scoped, but their DiT noise should not repeat every decode step; and unseeded requests should keep the fast batched random path.

**Comment by elismasilva** [comment]
> Thanks for the update — the API compatibility part looks good to me now.
> 
> I still have one concern about how the RNG is implemented for the DiT noise. The current approach creates a new `torch.Generator(...).manual_seed(seed)` inside each runner invocation. This avoids global RNG contamination, but it also means the same request receives the same initial DiT noise at every autoregressive decode step. In practice this can noticeably hurt generation quality, because each diffusion call starts from identical noise.
....

Thanks again for the follow-up review. I addressed the remaining DiT RNG concern you pointed out.

## 1. Per-request RNG stream now advances across decode steps

The seeded DiT noise path no longer recreates the same `torch.Generator(...).manual_seed(seed)` state on every runner invocation.

Instead, each `VoxCPM` and `VoxCPM2` sequence now carries a request-local `seed_step` counter. The engine passes both `seed` and `seed_step` into the runner payload, and `postprocess_seq()` increments `seed_step` after each successfully generated latent.

The runner derives a deterministic per-step seed with:

```python
derive_step_seed(seed, seed_step)
```

So the same original request seed still produces a reproducible RNG stream across runs, but each autoregressive DiT call receives different initial `z_noise` instead of restarting from identical noise.

## 2. Unseeded requests keep the fast batched path

I also separated the seeded and unseeded noise paths.

The common unseeded path now uses the original batched random generation pattern:

```python
torch.randn((bsz, feat_dim, patch_size), ...)
```

For mixed batches, the runner generates the full batch normally and then overwrites only the rows that belong to seeded requests. For fully seeded batches, it allocates the batch tensor and fills each row from the request-local derived seed without consuming the process-global RNG stream.

This keeps unseeded requests on the fast batched path while making seeded requests deterministic, request-scoped, and advanced across decode steps.

## Validation

Targeted unit tests:

```powershell
python -m pytest tests/unit/test_seed_utils.py tests/unit/test_voxcpm_engine_max_model_len.py::test_postprocess_advances_seed_step_after_generated_latent tests/unit/test_voxcpm2_engine_max_model_len.py::test_postprocess_advances_seed_step_after_generated_latent tests/unit/test_voxcpm2_chunk_sizes.py::test_voxcpm2_runner_slices_decoded_waveform_with_decoder_chunk_size -q
```

Focused regression set:

```powershell
python -m pytest tests/unit/test_seed_utils.py tests/unit/test_voxcpm_engine_max_model_len.py tests/unit/test_voxcpm2_engine_max_model_len.py tests/unit/test_voxcpm2_chunk_sizes.py -q
```

Result:

```text
18 passed
```

Syntax validation:

```powershell
python -m compileall nanovllm_voxcpm/utils/seed.py nanovllm_voxcpm/models/voxcpm/engine.py nanovllm_voxcpm/models/voxcpm/runner.py nanovllm_voxcpm/models/voxcpm2/engine.py nanovllm_voxcpm/models/voxcpm2/runner.py tests/unit/test_seed_utils.py tests/unit/test_voxcpm_engine_max_model_len.py tests/unit/test_voxcpm2_engine_max_model_len.py tests/unit/test_voxcpm2_chunk_sizes.py
```

Implementation commit:

```text
461c351 Fix per-step seeded DiT noise
```

**Comment by a710128** [comment]
Thanks for the latest update — this version looks good functionally to me. The seeded DiT noise now advances per step, the unseeded fast path is preserved, and the API compatibility issue looks addressed.

The only remaining thing I noticed is formatting/trailing whitespace: `git diff --check origin/main...HEAD` still reports several lines. Could you please run `black` or otherwise clean up the formatting before merge?

**Comment by elismasilva** [comment]
done! tk you!

**Review by a710128 — APPROVED** [review]
Thanks for the updates! This looks good to me now.

The seeded DiT noise now advances per step, the unseeded fast path is preserved, the deployment compatibility issue is addressed, and the formatting/trailing whitespace issues are cleaned up. I also re-ran `git diff --check` and a lightweight compile/seed-utils test locally, and they passed.
